### PR TITLE
Many fixes

### PR DESCRIPTION
**Pull request by kevmoo**

### CODE REVIEW

**Comment by kevmoo** [comment]
Follow-on from https://github.com/dart-lang/json_serializable/pull/13

File: lib/src/json_serializable_generator.dart [hunk 1]
```diff
@@ -221,14 +227,58 @@ class JsonSerializableGenerator
         // ...resetting `isList` to `false`.
         isList = false;
       }
-    }
 
-    if (!isList && _coreIterableChecker.isAssignableFromType(fieldType)) {
-      // Then we need to add `?.toList()
-      expression += "?.toList()";
+      if (!isList) {
+        // Then we need to add `?.toList()
```
Comment (natebosch) [inline lib/src/json_serializable_generator.dart]: avoid comments that only repeat what the code is doing - if you want to keep a comment here it should say why we need this.

File: lib/src/json_serializable_generator.dart [hunk 2]
```diff
@@ -221,14 +227,58 @@ class JsonSerializableGenerator
...
+            "$expression.values.map(($substitute) => $subFieldValue))";
+      }
+    } else if (fieldType.isDynamic ||
+        fieldType.isObject ||
+        _stringBoolNumChecker.isAssignableFromType(fieldType)) {
+      return expression;
     }
 
-    return expression;
+    return "$expression /* unsafe */";
```
Comment (natebosch) [inline lib/src/json_serializable_generator.dart]: This went from 46 lines to 80 lines - consider breaking out some composed methods

File: lib/src/json_serializable_generator.dart [hunk 3]
```diff
@@ -221,14 +227,58 @@ class JsonSerializableGenerator
...
+          return expression;
+        }
+
+        // convert
+
+        var substitute = "v$depth";
+        var subFieldValue =
+            _fieldToJsonMapValue(substitute, valueType, depth + 1);
+
+        // In this case, we're going to create a new Map with matching reified
```
Comment (natebosch) [inline lib/src/json_serializable_generator.dart]: more comments that only say what is happening...

File: lib/src/json_serializable_generator.dart [hunk 4]
```diff
@@ -221,14 +227,58 @@ class JsonSerializableGenerator
...
 
-    if (!isList && _coreIterableChecker.isAssignableFromType(fieldType)) {
-      // Then we need to add `?.toList()
-      expression += "?.toList()";
+      if (!isList) {
+        // Then we need to add `?.toList()
+        expression += "?.toList()";
+      }
+
+      return expression;
```
Comment (natebosch) [inline lib/src/json_serializable_generator.dart]: there are lots of early returns in this method - it's a little hard to keep track

File: lib/src/json_serializable_generator.dart [hunk 5]
```diff
@@ -221,14 +227,58 @@ class JsonSerializableGenerator
...
+      var keyArg = args.first;
+      var valueType = args.last;
+
+      // We're not going to handle converting key types at the moment
+      // So the only safe types for key are dynamic/Object/String
+      var safeKey = keyArg.isDynamic ||
+          keyArg.isObject ||
+          _coreStringChecker.isExactlyType(keyArg);
+
+      var safeValue = false;
```
Comment (natebosch) [inline lib/src/json_serializable_generator.dart]: [optional] more declarative is `safeValue = valueType.isDynamic || valueType.isObject || _stringBoolNumChecker.isAssignableFromType(valueType);`

File: lib/src/json_serializable_generator.dart [hunk 6]
```diff
@@ -221,14 +227,58 @@ class JsonSerializableGenerator
...
+      }
+
+      if (safeKey) {
+        if (safeValue) {
+          return expression;
+        }
+
+        // convert
+
+        var substitute = "v$depth";
```
Comment (natebosch) [inline lib/src/json_serializable_generator.dart]: [nit] mixing single and double quotes in this file.

File: lib/src/json_serializable_generator.dart (hunk 1 above)
Comment (kevmoo) [inline lib/src/json_serializable_generator.dart]: ack

File: lib/src/json_serializable_generator.dart (hunk 4 above)
Comment (kevmoo) [inline lib/src/json_serializable_generator.dart]: Yeah – can look at in a future PR. Added a note.

File: lib/src/json_serializable_generator.dart (hunk 5 above)
Comment (kevmoo) [inline lib/src/json_serializable_generator.dart]: Nice!

File: lib/src/json_serializable_generator.dart (hunk 6 above)
Comment (kevmoo) [inline lib/src/json_serializable_generator.dart]: Waiting for https://github.com/dart-lang/linter/commit/aae769b4ba6e71bdb5e3eca46fb511f42a9d9f77

File: lib/src/json_serializable_generator.dart (hunk 3 above)
Comment (kevmoo) [inline lib/src/json_serializable_generator.dart]: ack

File: lib/src/json_serializable_generator.dart (hunk 4 above)
Comment (kevmoo) [inline lib/src/json_serializable_generator.dart]: ...or just refactor!

File: lib/src/json_serializable_generator.dart (hunk 2 above)
Comment (kevmoo) [inline lib/src/json_serializable_generator.dart]: ack

**Comment by kevmoo** [comment]
@natebosch PTAL

**Pull request merged**